### PR TITLE
Add msteams

### PR DESCRIPTION
**Pull request by zhan9san**
Related to #3076

As the `new line` is important to Markdown format, a `nested template definition` `__text_alert_list_markdown` is added.

For simplicity, only `text` field is used to send message in [Message card](https://learn.microsoft.com/en-us/outlook/actionable-messages/message-card-reference).

The message format is similar to Discord's.

Card design can be checked in [Card Playground](https://messagecardplayground.azurewebsites.net/) with following json text.

```json
{
	"@type": "MessageCard",
	"@context": "https://schema.org/extensions",
	"themeColor": "8C1A1A",
	"title": "[FIRING:1] InstanceDown (v critical core)",
	"text": "\n\nAlerts Firing:\n\nLabels:\n - alertname = InstanceDown\n - env = build-service\n - environment = v\n - instance = https://example.com\n - job = blackbox\n - severity = critical\n - type = core\n\nAnnotations:\n - description = https://example.com of job blackbox has been down for more than 5 minutes.\n - summary = Instance https://example.com down\n\nSource: http://v:9090/graph?g0.expr=probe_success%7Btype%3D%22core%22%7D+%3D%3D+0&g0.tab=1\n"
}
```

### CODE REVIEW

**Review by kbudde — CHANGES_REQUESTED** [review]

**Comment by kbudde** [comment]
Hey @zhan9san,
LGTM, just change the default in the docs (or change the default in code).

File: docs/configuration.md [hunk 1]
```diff
@@ -1044,6 +1046,27 @@ attributes:
 [ role_arn: <string> ]
 ```
 
+### `<teams_config>`
+
+Teams notifications are sent via the [Incoming Webhooks](https://learn.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/what-are-webhooks-and-connectors).
+
+```yaml
+# Whether to notify about resolved alerts.
+[ send_resolved: <boolean> | default = true ]
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply docs/configuration.md:1055]: @kbudde 

Thanks for your review. I think we need to keep the default value to `true`.

I have updated the default value in code.

**Review by kbudde — APPROVED** [review]

**Comment by zhan9san** [comment]
@simonpasquier 

Could you help review this PR?

File: notify/teams/teams.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+	var payload bytes.Buffer
+	if err = json.NewEncoder(&payload).Encode(t); err != nil {
+		return false, err
+	}
+
+	resp, err := notify.PostJSON(ctx, n.client, n.webhookURL.String(), &payload)
+	if err != nil {
+		return true, notify.RedactURL(err)
+	}
```
Comment (gotjosh) [inline notify/teams/teams.go:123]:
```suggestion
	}
	defer notify.Drain(resp)
```

Let's make sure we close the connection.

File: notify/teams/teams_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/notify/test"
+	"github.com/prometheus/alertmanager/types"
+)
+
+// This is a test URL that has been modified to not be valid.
+var testWebhookURL, _ = url.Parse("https://example.webhook.office.com/webhookb2/xxxxxx/IncomingWebhook/xxx/xxx")
+
+func TestTeamsRetry(t *testing.T) {
```
Comment (gotjosh) [inline notify/teams/teams_test.go:40]: Can we please add 1 more test that is similar to what we have in https://github.com/prometheus/alertmanager/pull/3252/files

I think it's important that we test:

1. The expected format of the JSON on the webhook api
2. the reason codes

File: notify/teams/teams.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return false, err
+	}
+
+	color := colorGrey
+	if alerts.Status() == model.AlertFiring {
+		color = colorRed
+	}
+	if alerts.Status() == model.AlertResolved {
+		color = colorGreen
+	}
```
Comment (simonpasquier) [inline notify/teams/teams.go:105]: (nit)
```suggestion
	switch alerts.Status() {
	case model.AlertFiring:
		color = colorRed
	case alerts.Status():
		color = colorGreen
	}
```

File: notify/teams/teams.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	key, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	level.Debug(n.logger).Log("incident", key)
+
+	alerts := types.Alerts(as...)
```
Comment (simonpasquier) [inline notify/teams/teams.go:83]: (nit) can you  move it right before L99?

**Comment by zhan9san** [comment]
Thanks for your careful review.

I just get back from a five-day vacation and I'll handle it later.

**Comment by TheMeier** [comment]
@zhan9san thanks a lot four you great contribution. Since the MessageCards are marked "legacy" I was wondering if it might be beneficial to explore switching to AdaptiveCards. Of cause only if that's not a lot of effort or causes more delays in this feature being available.

**Comment by zhan9san** [comment]
@TheMeier 

Thanks for pointing it out. I thought about the legacy issue at first.

But to make the PR simple, I just make a lift-shift migration to make MSTeams work.

Besides, this can be considered as an improvement in next PR.

File: cmd/alertmanager/main.go [hunk 6]
```diff
@@ -56,6 +56,7 @@ import (
 	"github.com/prometheus/alertmanager/notify/pushover"
 	"github.com/prometheus/alertmanager/notify/slack"
 	"github.com/prometheus/alertmanager/notify/sns"
+	"github.com/prometheus/alertmanager/notify/teams"
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply cmd/alertmanager/main.go]: The package is renamed to msteams in new commit

File: notify/teams/teams.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+type Notifier struct {
+	conf       *config.TeamsConfig
+	tmpl       *template.Template
+	logger     log.Logger
+	client     *http.Client
+	retrier    *notify.Retrier
+	webhookURL *config.SecretURL
+}
+
+// Message card https://learn.microsoft.com/en-us/outlook/actionable-messages/message-card-reference
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply notify/teams/teams.go]: It is fixed in new commit

File: notify/teams/teams.go (hunk 2 above)
Comment (zhan9san) [reply notify/teams/teams.go]: It is fixed in new commit

File: notify/teams/teams.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	resp, err := notify.PostJSON(ctx, n.client, n.webhookURL.String(), &payload)
+	if err != nil {
+		return true, notify.RedactURL(err)
+	}
+
+	shouldRetry, err := n.retrier.Check(resp.StatusCode, resp.Body)
+	if err != nil {
+		return shouldRetry, err
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply notify/teams/teams.go]: The 429 status code is not set in a standard way. Instead, if it hits rate limit, 200 OK is always send with 'Microsoft Teams endpoint returned HTTP error 429' in response content.

So `GetFailureReasonFromStatusCode` is renamed to `GetFailureReason` and add a new parameter `responseContent string` to address this issue

File: notify/teams/teams_test.go (hunk 3 above)
Comment (zhan9san) [reply notify/teams/teams_test.go]: The test is added in new commit

**Comment by zhan9san** [comment]
@simonpasquier  and @gotjosh 

Would you kindly review this PR?

File: notify/teams/teams.go (hunk 4 above)
Comment (zhan9san) [reply notify/teams/teams.go]: It is addressed in new commit

File: notify/teams/teams.go (hunk 5 above)
Comment (zhan9san) [reply notify/teams/teams.go]: It is addressed in new commit

**Comment by zhan9san** [comment]
Thank you again for taking the time to review the changes.
I just wanted to check in and see if there's anything that's unclear or needs further explanation.

Please let me know if there's anything I can do to make the process easier for you.

File: notify/msteams/msteams.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		return false, err
+	}
+
+	resp, err := n.postJSONFunc(ctx, n.client, n.webhookURL.String(), &payload)
+	if err != nil {
+		return true, notify.RedactURL(err)
+	}
+	defer notify.Drain(resp)
+
+	// https://learn.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/connectors-using?tabs=cURL#rate-limiting-for-connectors
```
Comment (simonpasquier) [inline notify/msteams/msteams.go:131]: Hmm so a MSTeams webhook doesn't reply with 429 when rate-limiting and the only option is to check for the response body? 
As it is significantly different from what other receivers do and to limit the impact on the rest of the code, I'd prefer to defer the retry handling to a follow-up PR. I'm not quite comfortable merging the changes to  notify/util.go right now.

Comment (zhan9san) [reply notify/msteams/msteams.go:131]: Thanks for your attention.

> so a MSTeams webhook doesn't reply with 429 when rate-limiting and the only option is to check for the response body?

Yes, it's not be done in a standard way. As we can see in the MSTeams official document, it is the only option.

I'll delete the retry handling in this PR.

File: docs/configuration.md [hunk 10]
```diff
@@ -717,6 +719,27 @@ tls_config:
...
 
+### `<msteams_config>`
+
+Microsoft Teams notifications are sent via the [Incoming Webhooks](https://learn.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/what-are-webhooks-and-connectors) API endpoint.
+
+```yaml
+# Whether to notify about resolved alerts.
+[ send_resolved: <boolean> | default = true ]
+
+# The Incoming webhook URL.
```
Comment (simonpasquier) [inline docs/configuration.md:730]: (nit)
```suggestion
# The incoming webhook URL.
```

**Review by simonpasquier — COMMENTED** [review]
Sorry for the delay! Can you add `msteams` to the list of integrations here?
https://github.com/prometheus/alertmanager/blob/83304da5070d8a12642b7a027d4c59060813c0e3/notify/notify.go#L283-L294

Otherwise this is good for me.

**Review by simonpasquier — APPROVED** [review]
Thanks! I'll wait a few days before merging in case @gotjosh has more comments :)

**Review by gotjosh — APPROVED** [review]
LGTM

**Review by gotjosh — APPROVED** [review]
LGTM

**Comment by gotjosh** [comment]
Thank you very much for your patience and your contributions ❤️

**Comment by rgarcia89** [comment]
Now all we need is a new release to be build 😁

**Comment by SergkeiM** [comment]
@zhan9san Amazing thanks a lot for this, we had to use Telegram until MS Teams is added, now we can use 1 messenger only :)

@rgarcia89 I hope will be soon, finger crossed

**Comment by rgarcia89** [comment]
@gotjosh @simonpasquier any idea when we can count with a new release? It looks like there is quite some demand for this PR go become available 🤔

**Comment by rgarcia89** [comment]
Anyone tried to build an image including the merge by itself? Is there anything I should give a special look? I think this is the only reasonable way to get an updated image in time

**Comment by roidelapluie** [comment]
You can use the :main image.

**Comment by SergkeiM** [comment]
Hi @gotjosh this was added to 0.26 release? I don't see this in changelog. Thank you.

**Comment by gotjosh** [comment]
Dear all - this as now been released as part of 0.26

**Comment by SergkeiM** [comment]
Hi @gotjosh Thank you, already changed it in our Ceph Cluster:
image 🚀

haveing issue in Prometheus though: (3724)[https://github.com/prometheus-community/helm-charts/issues/3724]

File: config/config.go [hunk 11]
```diff
@@ -896,6 +907,7 @@ type Receiver struct {
 	SNSConfigs       []*SNSConfig       `yaml:"sns_configs,omitempty" json:"sns_configs,omitempty"`
 	TelegramConfigs  []*TelegramConfig  `yaml:"telegram_configs,omitempty" json:"telegram_configs,omitempty"`
 	WebexConfigs     []*WebexConfig     `yaml:"webex_configs,omitempty" json:"webex_configs,omitempty"`
+	MSTeamsConfigs   []*MSTeamsConfig   `yaml:"msteams_configs,omitempty" json:"teams_configs,omitempty"`
```
Comment (yeya24) [inline config/config.go:910]: Should we align the configuration name? I am wondering why we call `msteams` for yaml but `teams` for json.
@zhan9san @simonpasquier

Comment (zhan9san) [reply config/config.go:910]: @yeya24 

Thanks for pointing it out. It has to be a typo.

I'll submit a PR to fix it.

**Pull request merged**